src/gridfs/download.rs

```diff
@@ -0,0 +1,181 @@
+                }
+                .into());
+            }
+        };
+
+        self.download_to_tokio_writer_common(file, destination)
+            .await
+    }
+
+    /// Downloads the contents of the stored file specified by `filename` and writes the contents to
+    /// the `destination`. If there are multiple files with the same filename, the `revision` in the
+    /// options provided is used to determine which one to download. If no `revision` is specified,
+    /// the most recent file with the given filename is chosen.
+    pub async fn download_to_futures_0_3_writer_by_name<T>(
+        &self,
+        filename: impl AsRef<str>,
+        destination: &mut T,
+        options: impl Into<Option<GridFsDownloadByNameOptions>>,
+    ) -> Result<()>
+    where
+        T: futures_util::io::AsyncWrite + std::marker::Unpin,
+    {
+        self.download_to_tokio_writer_by_name(filename, &mut destination.compat_write(), options)
+            .await
+    }
+
+    async fn download_to_tokio_writer_common<T>(
+        &self,
+        file: FilesCollectionDocument,
+        destination: &mut T,
+    ) -> Result<()>
+    where
+        T: tokio::io::AsyncWrite + std::marker::Unpin,
+    {
+        let total_bytes_expected = file.length;
+        let chunk_size = file.chunk_size as u64;
+
+        if total_bytes_expected == 0 {
+            return Ok(());
+        }
+
+        let options = FindOptions::builder()
+            .sort(doc! { "n": 1 })
+            .read_concern(self.read_concern().cloned())
+            .selection_criteria(self.selection_criteria().cloned())
+            .build();
+        let mut cursor = self
+            .inner
+            .chunks
+            .find(doc! { "files_id": &file.id }, options)
+            .await?;
+
+        let mut n = 0;
+        while cursor.advance().await? {
+            let chunk = cursor.deserialize_current()?;
+            if chunk.n != n {
+                return Err(ErrorKind::GridFS {
+                    message: format!("missing chunk {} in file", n),
+                }
+                .into());
+            }
+
+            let chunk_length = chunk.data.bytes.len();
+            let expected_length =
+                std::cmp::min(total_bytes_expected - chunk_size * n as u64, chunk_size);
+            if chunk_length as u64 != expected_length {
+                return Err(ErrorKind::GridFS {
+                    message: format!(
+                        "chunk {} was expected to be {} bytes but is {} bytes",
+                        n, chunk_length, expected_length
+                    ),
+                }
+                .into());
+            }
+
+            destination.write_all(&chunk.data.bytes).await?;
+            n += 1;
+        }
+
+        let expected_n =
+            total_bytes_expected / chunk_size + u64::from(total_bytes_expected % chunk_size != 0);
+        if (n as u64) < expected_n {
```

Comment:
Would we want to make this more general and check for `expected_n != n`? Then we could say "number of chunks read (n) did not match expected number of chunks (expected_n)".